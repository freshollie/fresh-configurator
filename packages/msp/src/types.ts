import SerialPort from "@serialport/stream";
import { MspParser } from "./parser";
import WriteBuffer from "./writebuffer";

export type MspInfo = {
  mspProtocolVersion: number;
  apiVersion: string;
};

export type ExecutionLocks = Record<number, Promise<void>>;

export type MspRequest = {
  close: () => void;
  response: Promise<ArrayBuffer>;
};

export type Connection = {
  serial: SerialPort;
  parser: MspParser;
  requests: Record<string, MspRequest>;
  bytesWritten: number;
  bytesRead: number;
  packetErrors: number;
  mspInfo: MspInfo;
};

export type ConnectionOptions = {
  baudRate?: number;
};

export type OnCloseCallback = () => void;

export type OpenConnectionFunction = {
  (
    port: string,
    options?: ConnectionOptions,
    onClose?: OnCloseCallback
  ): Promise<void>;
  (port: string, onClose?: OnCloseCallback): Promise<void>;
};

export type MspCommand = {
  code: number;
  data?: WriteBuffer | Buffer;
  timeout?: number;
};
