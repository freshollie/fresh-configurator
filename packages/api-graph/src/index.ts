import { makeExecutableSchema } from "@graphql-tools/schema";
import { startTicks } from "./mock/api";
import graph from "./graph";
import context, { mockedContext } from "./context";

export const schema = makeExecutableSchema(graph);
export {
  mockedContext as mockedDeviceContext,
  context,
  startTicks as startMockDevice,
};
