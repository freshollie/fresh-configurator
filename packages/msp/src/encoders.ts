import { crc8DvbS2Data } from "./utils";

export const encodeMessageV1 = (code: number, data?: Buffer): Buffer => {
  let bufferOut;
  // always reserve 6 bytes for protocol overhead !
  if (data) {
    const size = data.length + 6;
    let checksum = 0;

    bufferOut = new ArrayBuffer(size);
    const bufView = new Uint8Array(bufferOut);

    bufView[0] = 36; // $
    bufView[1] = 77; // M
    bufView[2] = 60; // <
    bufView[3] = data.length;
    bufView[4] = code;

    checksum = bufView[3] ^ bufView[4];

    for (let i = 0; i < data.length; i += 1) {
      bufView[i + 5] = data[i]!;

      checksum ^= bufView[i + 5]!;
    }

    bufView[5 + data.length] = checksum;
  } else {
    bufferOut = new ArrayBuffer(6);
    const bufView = new Uint8Array(bufferOut);

    bufView[0] = 36; // $
    bufView[1] = 77; // M
    bufView[2] = 60; // <
    bufView[3] = 0; // data length
    bufView[4] = code; // code
    bufView[5] = bufView[3] ^ bufView[4]; // checksum
  }
  return Buffer.from(bufferOut);
};

export const encodeMessageV2 = (code: number, data?: Buffer): Buffer => {
  const dataLength = data ? data.length : 0;
  // 9 bytes for protocol overhead
  const bufferSize = dataLength + 9;
  const bufferOut = new ArrayBuffer(bufferSize);
  const bufView = new Uint8Array(bufferOut);
  bufView[0] = 36; // $
  bufView[1] = 88; // X
  bufView[2] = 60; // <
  bufView[3] = 0; // flag
  bufView[4] = code & 0xff;
  bufView[5] = (code >> 8) & 0xff;
  bufView[6] = dataLength & 0xff;
  bufView[7] = (dataLength >> 8) & 0xff;
  for (let i = 0; i < dataLength; i += 1) {
    bufView[8 + i] = data![i]!;
  }
  bufView[bufferSize - 1] = crc8DvbS2Data(bufView, 3, bufferSize - 1);
  return Buffer.from(bufferOut);
};
