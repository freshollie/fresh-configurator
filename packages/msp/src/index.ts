/* eslint-disable import/first */
export {
  ports,
  connections,
  isOpen,
  open,
  close,
  bytesRead,
  bytesWritten,
  packetErrors,
  apiVersion,
  execute,
  baudRate,
  initialiseSerialBackend,
} from "./connection";
export { default as WriteBuffer } from "./writebuffer";
export { default as MspDataView } from "./dataview";
