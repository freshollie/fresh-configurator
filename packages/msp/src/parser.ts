/**
 * The MSP Parser
 *
 * This module is designed to parse bytes and
 * emit data once the full request has been
 * received, parsed, and checked.
 *
 * The sourcecode in this module can mostly be attributed
 * to the betaflight-configurator repository, but has
 * been ported to typescript and redesigned to work
 * with node-serialport
 */

import { Transform } from "stream";
import debug from "debug";
import { crc8DvbS2 } from "./utils";

const log = debug("msp:parser");

const Symbols = {
  BEGIN: "$".charCodeAt(0),
  PROTO_V1: "M".charCodeAt(0),
  PROTO_V2: "X".charCodeAt(0),
  FROM_MWC: ">".charCodeAt(0),
  TO_MWC: "<".charCodeAt(0),
  UNSUPPORTED: "!".charCodeAt(0),
};

enum Constants {
  PROTOCOL_V1 = 1,
  PROTOCOL_V2 = 2,
  JUMBO_FRAME_MIN_SIZE = 255,
}

enum DecoderStates {
  IDLE = 0,
  PROTO_IDENTIFIER = 1,
  DIRECTION_V1 = 2,
  DIRECTION_V2 = 3,
  FLAG_V2 = 4,
  PAYLOAD_LENGTH_V1 = 5,
  PAYLOAD_LENGTH_JUMBO_LOW = 6,
  PAYLOAD_LENGTH_JUMBO_HIGH = 7,
  PAYLOAD_LENGTH_V2_LOW = 8,
  PAYLOAD_LENGTH_V2_HIGH = 9,
  CODE_V1 = 10,
  CODE_JUMBO_V1 = 11,
  CODE_V2_LOW = 12,
  CODE_V2_HIGH = 13,
  PAYLOAD_V1 = 14,
  PAYLOAD_V2 = 15,
  CHECKSUM_V1 = 16,
  CHECKSUM_V2 = 17,
}

export type MspMessage = {
  code: number;
  data: ArrayBuffer;
  crcError: boolean;
  unsupported: number;
  direction: number;
};

export class MspParser extends Transform {
  private state: DecoderStates;

  private messageDirection: number;

  private code: number;

  private message_length_expected: number;

  private message_length_received: number;

  private message_buffer: ArrayBuffer;

  private message_buffer_uint8_view: Uint8Array;

  private message_checksum: number;

  private crcError: boolean;

  private unsupported: number;

  constructor() {
    super();

    this.state = 0;
    this.messageDirection = 1;
    this.code = 0;
    this.message_length_expected = 0;
    this.message_length_received = 0;
    this.message_buffer = new ArrayBuffer(0);
    this.message_buffer_uint8_view = new Uint8Array(0);
    this.message_checksum = 0;
    this.crcError = false;
    this.unsupported = 0;
  }

  // eslint-disable-next-line no-underscore-dangle
  public _transform(
    chunk: Buffer,
    _: string,
    cb: (error?: Error | null, data?: Buffer) => void
  ): void {
    log(`Received data ${chunk.toJSON().data}`);
    const data = new Uint8Array(chunk);

    data.forEach((byte) => {
      switch (this.state) {
        case DecoderStates.IDLE: // sync char 1
          if (byte === Symbols.BEGIN) {
            this.state = DecoderStates.PROTO_IDENTIFIER;
          }
          break;
        case DecoderStates.PROTO_IDENTIFIER: // sync char 2
          switch (byte) {
            case Symbols.PROTO_V1:
              this.state = DecoderStates.DIRECTION_V1;
              break;
            case Symbols.PROTO_V2:
              this.state = DecoderStates.DIRECTION_V2;
              break;
            default:
              log(`Unknown protocol char ${String.fromCharCode(byte)}`);
              this.state = DecoderStates.IDLE;
          }
          break;
        case DecoderStates.DIRECTION_V1: // direction (should be >)
        case DecoderStates.DIRECTION_V2:
          this.unsupported = 0;
          switch (byte) {
            case Symbols.FROM_MWC:
              this.messageDirection = 1;
              break;
            case Symbols.TO_MWC:
              this.messageDirection = 0;
              break;
            case Symbols.UNSUPPORTED:
              this.unsupported = 1;
              break;
            default:
          }
          this.state =
            this.state === DecoderStates.DIRECTION_V1
              ? DecoderStates.PAYLOAD_LENGTH_V1
              : DecoderStates.FLAG_V2;
          break;
        case DecoderStates.FLAG_V2:
          // Ignored for now
          this.state = DecoderStates.CODE_V2_LOW;
          break;
        case DecoderStates.PAYLOAD_LENGTH_V1:
          this.message_length_expected = byte;

          if (this.message_length_expected === Constants.JUMBO_FRAME_MIN_SIZE) {
            this.state = DecoderStates.CODE_JUMBO_V1;
          } else {
            this.initializeReadBuffer();
            this.state = DecoderStates.CODE_V1;
          }

          break;
        case DecoderStates.PAYLOAD_LENGTH_V2_LOW:
          this.message_length_expected = byte;
          this.state = DecoderStates.PAYLOAD_LENGTH_V2_HIGH;
          break;
        case DecoderStates.PAYLOAD_LENGTH_V2_HIGH:
          this.message_length_expected |= byte << 8;
          this.initializeReadBuffer();
          this.state =
            this.message_length_expected > 0
              ? DecoderStates.PAYLOAD_V2
              : DecoderStates.CHECKSUM_V2;
          break;
        case DecoderStates.CODE_V1:
        case DecoderStates.CODE_JUMBO_V1:
          this.code = byte;
          if (this.message_length_expected > 0) {
            // process payload
            if (this.state === DecoderStates.CODE_JUMBO_V1) {
              this.state = DecoderStates.PAYLOAD_LENGTH_JUMBO_LOW;
            } else {
              this.state = DecoderStates.PAYLOAD_V1;
            }
          } else {
            // no payload
            this.state = DecoderStates.CHECKSUM_V1;
          }
          break;
        case DecoderStates.CODE_V2_LOW:
          this.code = byte;
          this.state = DecoderStates.CODE_V2_HIGH;
          break;
        case DecoderStates.CODE_V2_HIGH:
          this.code |= byte << 8;
          this.state = DecoderStates.PAYLOAD_LENGTH_V2_LOW;
          break;
        case DecoderStates.PAYLOAD_LENGTH_JUMBO_LOW:
          this.message_length_expected = byte;
          this.state = DecoderStates.PAYLOAD_LENGTH_JUMBO_HIGH;
          break;
        case DecoderStates.PAYLOAD_LENGTH_JUMBO_HIGH:
          this.message_length_expected |= byte << 8;
          this.initializeReadBuffer();
          this.state = DecoderStates.PAYLOAD_V1;
          break;
        case DecoderStates.PAYLOAD_V1:
        case DecoderStates.PAYLOAD_V2:
          this.message_buffer_uint8_view[this.message_length_received] = byte;
          this.message_length_received += 1;

          if (this.message_length_received >= this.message_length_expected) {
            this.state =
              this.state === DecoderStates.PAYLOAD_V1
                ? DecoderStates.CHECKSUM_V1
                : DecoderStates.CHECKSUM_V2;
          }
          break;
        case DecoderStates.CHECKSUM_V1:
          if (this.message_length_expected >= Constants.JUMBO_FRAME_MIN_SIZE) {
            this.message_checksum = Constants.JUMBO_FRAME_MIN_SIZE;
          } else {
            this.message_checksum = this.message_length_expected;
          }
          this.message_checksum ^= this.code;
          if (this.message_length_expected >= Constants.JUMBO_FRAME_MIN_SIZE) {
            this.message_checksum ^= this.message_length_expected & 0xff;
            this.message_checksum ^=
              (this.message_length_expected & 0xff00) >> 8;
          }

          this.message_buffer_uint8_view
            .slice(0, this.message_length_received)
            .forEach((messageByte) => {
              this.message_checksum ^= messageByte;
            });

          this.dispatchMessage(byte);
          break;
        case DecoderStates.CHECKSUM_V2:
          this.message_checksum = 0;
          this.message_checksum = crc8DvbS2(this.message_checksum, 0); // flag
          this.message_checksum = crc8DvbS2(
            this.message_checksum,
            this.code & 0xff
          );
          this.message_checksum = crc8DvbS2(
            this.message_checksum,
            (this.code & 0xff00) >> 8
          );
          this.message_checksum = crc8DvbS2(
            this.message_checksum,
            this.message_length_expected & 0xff
          );
          this.message_checksum = crc8DvbS2(
            this.message_checksum,
            (this.message_length_expected & 0xff00) >> 8
          );

          this.message_buffer_uint8_view
            .slice(0, this.message_length_received)
            .forEach((messageByte) => {
              this.message_checksum = crc8DvbS2(
                this.message_checksum,
                messageByte
              );
            });
          this.dispatchMessage(byte);
          break;
        default:
          log(`Unknown state detected: ${this.state}`);
      }
    });
    cb();
  }

  private initializeReadBuffer(): void {
    this.message_buffer = new ArrayBuffer(this.message_length_expected);
    this.message_buffer_uint8_view = new Uint8Array(this.message_buffer);
  }

  private dispatchMessage(expectedChecksum: number): void {
    let data: ArrayBuffer;
    if (this.message_checksum === expectedChecksum) {
      // message received, store dataview
      data = new DataView(this.message_buffer, 0, this.message_length_expected)
        .buffer;
    } else {
      log(
        `code: ${this.code} - crc failed, expectedChecksum=${expectedChecksum}, checksum=${this.message_checksum}`
      );
      this.crcError = true;
      data = new ArrayBuffer(0);
    }
    log(`Parsed message: ${Buffer.from(data).toJSON().data}`);

    this.emit("data", {
      code: this.code,
      data,
      crcError: this.crcError,
      unsupported: this.unsupported,
      direction: this.messageDirection,
    });
    this.reset();
  }

  private reset(): void {
    this.message_length_received = 0;
    this.state = DecoderStates.IDLE;
    this.crcError = false;
  }

  // eslint-disable-next-line no-underscore-dangle
  public _flush(cb: () => void): void {
    this.reset();
    cb();
  }
}
