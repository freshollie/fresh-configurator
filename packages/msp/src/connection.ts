/**
 * MSP Serial connections. This allows us to make multiple
 * serial connections, and enables execution of commands
 * on the given connection.
 *
 * The module handles making requests, and dealing with only
 * making one of the same request at a time.
 */

import SerialPort from "@serialport/stream";
import debug from "debug";
import MspDataView from "./dataview";
import { MspMessage, MspParser } from "./parser";
import { encodeMessageV2, encodeMessageV1 } from "./encoders";

import {
  Connection,
  MspCommand,
  ConnectionOptions,
  OpenConnectionFunction,
  OnCloseCallback,
} from "./types";

// Import bindings when we actually need them, so that libraries
// can import this library without the bindings needing to be available
// at import time
const initialiseBindings = async (): Promise<void> => {
  const { default: Binding } = await import("@serialport/bindings");
  SerialPort.Binding = Binding;
};

const log = debug("msp:connection");

const connectionsMap: Record<string, Connection> = {};

/**
 * Execute the given MspCommand on the given port.
 *
 * Return the data received from the command
 */
export const execute = async (
  port: string,
  { code, data, timeout = 2500 }: MspCommand
): Promise<MspDataView> => {
  const connection = connectionsMap[port];
  if (!connection) {
    throw new Error(`${port} is not open`);
  }

  const sendData = data ? Buffer.from(Array.from(data)) : undefined;

  const { parser, serial, requests } = connection;

  const request =
    code > 254
      ? encodeMessageV2(code, sendData)
      : encodeMessageV1(code, sendData);

  const requestKey = request.toString("utf-8");
  let mspRequest = requests[requestKey];

  // Prevent FC lockup by checking if this request is currently being performed
  // and if not, making the request
  if (mspRequest === undefined) {
    let onError: (err: Error) => void | undefined;
    const response = new Promise<ArrayBuffer>((resolve, reject) => {
      // Throw an error if timeout is reached
      const timeoutId = setTimeout(() => {
        delete requests[requestKey];
        // eslint-disable-next-line @typescript-eslint/no-use-before-define
        parser.off("data", onData);
        reject(new Error("Timed out during execution"));
      }, timeout);

      onError = (e) => {
        reject(e);
        clearTimeout(timeoutId);
        delete requests[requestKey];
      };

      function onData(message: MspMessage): void {
        if (message.code === code) {
          if (message.crcError && connection) {
            connection.packetErrors += 1;
            return;
          }
          log(
            `request: ${request.toJSON().data}, response: ${
              Buffer.from(message.data).toJSON().data
            }, byteLength: ${message.data.byteLength}`
          );

          delete requests[requestKey];
          clearTimeout(timeoutId);

          resolve(message.data);
          parser.off("data", onData);
        }
      }

      parser.on("data", onData);
    });

    mspRequest = {
      response,
      close: () => onError(new Error("Connection closed")),
    };
    requests[requestKey] = mspRequest;

    log(`Writing ${request.toJSON().data} to ${port}`);
    serial.write(request, (error) => {
      if (error) {
        log(
          `Error writing to serial port ${error.message}, rejecting execution`
        );
        onError(new Error(`Could not write to serial port: ${error.message}`));
      } else {
        connection.bytesWritten += Buffer.byteLength(request);
      }
    });
  }

  // make every DataView unique to each request, even though
  // they are accessing the same set of data
  return mspRequest.response.then(
    (responseData) => new MspDataView(responseData)
  );
};

export const ports = async (): Promise<string[]> => {
  // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
  if (!SerialPort.Binding) {
    await initialiseBindings();
  }
  return SerialPort.list().then((data) => data.map(({ path }) => path));
};

/**
 * Close the given port
 */
export const close = async (port: string): Promise<void> => {
  const connection = connectionsMap[port];
  if (!connection) {
    return;
  }
  const { serial } = connection;
  delete connectionsMap[port];

  const closePromise = serial.isOpen
    ? new Promise((resolve) => serial.on("close", resolve))
    : Promise.resolve();
  await new Promise((resolve) => serial.close(resolve));

  Object.values(connection.requests).forEach((request) => {
    request.close();
  });

  await closePromise;
};

/**
 * Open a serial connection on the given
 * port, onClose will be emited once the
 * connection is closed
 */
export const open: OpenConnectionFunction = async (
  port: string,
  options?: ConnectionOptions | OnCloseCallback,
  onClose?: OnCloseCallback
) => {
  if (connectionsMap[port]) {
    throw new Error(`${port} is already open`);
  }
  log(`opening ${port}`);

  // Handle overloads
  let onCloseCallback = onClose;
  let connectionOptions = options;

  if (typeof connectionOptions === "function") {
    onCloseCallback = connectionOptions;
    connectionOptions = {};
  }

  // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
  if (!SerialPort.Binding) {
    await initialiseBindings();
  }

  const serial = new SerialPort(port, {
    baudRate: 115200,
    ...connectionOptions,
    autoOpen: false,
  });

  await new Promise<void>((resolve, reject) => {
    serial.open((err) => {
      if (err) {
        log(`error opening ${port}`);
        reject(err);
      } else {
        resolve();
      }
    });
  });

  const parser = serial.pipe(new MspParser());
  parser.setMaxListeners(1000000);

  const connection = {
    serial,
    parser,
    requests: {},
    bytesRead: 0,
    bytesWritten: 0,
    packetErrors: 0,
    mspInfo: {
      apiVersion: "0",
      mspProtocolVersion: 0,
    },
  };

  connectionsMap[port] = connection;

  log(`reading API version from ${port}`);
  try {
    const response = await execute(port, { code: 1 });
    connection.mspInfo = {
      mspProtocolVersion: response.readU8(),
      apiVersion: `${response.readU8()}.${response.readU8()}.0`,
    };

    log(`read apiVersion=${connection.mspInfo.apiVersion}`);
  } catch (e) {
    log(`failed to read MSP info from ${port}, failed to open`);
    await close(port);
    throw new Error(`Could not read MSP version from ${port}`);
  }

  // Disconnect isn't notified unless we write to the serial
  // port after it disconnects. This interval constantly checks
  // the available ports, and issues a disconnect if the port
  // is removed
  const disconnectMonitor = setInterval(async () => {
    const list = await ports();
    if (!list.includes(port) && serial.isOpen) {
      close(port);
    }
  }, 250);

  // valid connection, setup listeners
  serial.on("data", (data: Buffer) => {
    connection.bytesRead += Buffer.byteLength(data);
  });

  serial.on("error", (e) => {
    log(`${port} on error received: ${e?.message}`);
    // Don't trigger close if this serial port is already
    // closed, as it means that the closing
    // will be dealt with by the close event
    if (serial.isOpen) {
      close(port);
    }
  });

  serial.on("close", async () => {
    log(`${port} on close received`);
    clearInterval(disconnectMonitor);
    await close(port);
    onCloseCallback?.();
  });

  connection.bytesRead = 0;
  connection.bytesWritten = 0;
  connection.packetErrors = 0;
};

export const connections = (): string[] => Object.keys(connectionsMap);
export const isOpen = (port: string): boolean => !!connectionsMap[port];

export const bytesRead = (port: string): number =>
  connectionsMap[port]?.bytesRead ?? 0;

export const bytesWritten = (port: string): number =>
  connectionsMap[port]?.bytesWritten ?? 0;

export const packetErrors = (port: string): number =>
  connectionsMap[port]?.packetErrors ?? 0;

export const apiVersion = (port: string): string =>
  connectionsMap[port]?.mspInfo.apiVersion ?? "0.0.0";

/**
 * Private, used for testing
 */
export const raw = (port: string): SerialPort | undefined =>
  connectionsMap[port]?.serial;

/**
 * Private, used for testing
 */
export const reset = async (): Promise<void> => {
  await Promise.all(Object.keys(connectionsMap).map((port) => close(port)));
};
