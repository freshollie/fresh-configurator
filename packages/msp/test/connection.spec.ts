import MockBinding from "@serialport/binding-mock";
import binding from "@serialport/bindings";
import SerialPort from "@serialport/stream";
import flushPromises from "flush-promises";
import { raw, reset, execute, apiVersion } from "../src/connection";
import {
  open,
  connections,
  isOpen,
  ports,
  close,
  bytesRead,
  bytesWritten,
  packetErrors,
} from "../src";

import { encodeMessageV1, encodeMessageV2 } from "../src/encoders";

const mockMspDevices = ["/dev/something", "/dev/somethingelse"];
const mockPorts = ["/dev/non-msp-device", ...mockMspDevices];

const mspInfoRequest = encodeMessageV1(1);

/**
 * Get the data written to the given port
 * after the msp info request was written
 */
const writtenData = (port: string): Buffer =>
  (raw(port)?.binding as MockBinding).recording.slice(
    mspInfoRequest.byteLength
  );

const reply = (port: string, data: Buffer): void => {
  (raw(port)?.binding as MockBinding).emitData(data);
};

const realSetTimeout = setTimeout;

let killMspReplying = false;
/**
 * Automatically reply MSP info for the given port
 * when it's opened
 */
const handleMspInfoReply = async (port: string): Promise<void> => {
  let currentPort = raw(port);
  while (!killMspReplying) {
    try {
      if (
        currentPort !== raw(port) &&
        (raw(port)?.binding as MockBinding).recording.compare(
          mspInfoRequest
        ) === 0
      ) {
        currentPort = raw(port);
        // Reply with some mock MSP info
        reply(port, Buffer.from([36, 77, 62, 3, 1, 0, 1, 40, 43]));
      }
    } catch (e) {}
    // Respond to an open request within 100ms
    await new Promise((resolve) => realSetTimeout(resolve, 10));
  }
};

mockMspDevices.forEach((port) => handleMspInfoReply(port));

afterAll(() => {
  killMspReplying = true;
});

beforeEach(() => {
  // Reset the bindings every iteration to ensure
  // that every function can start from an empty state
  SerialPort.Binding = undefined as any;
  mockPorts.forEach((path) => {
    MockBinding.createPort(path, {
      record: true,
    });
  });
});

afterEach(async () => {
  jest.clearAllTimers();
  jest.useRealTimers();
  await reset();
  MockBinding.reset();
});

describe("open", () => {
  it("should open an MSP connection to the given port", async () => {
    await open("/dev/something");
    expect(isOpen("/dev/something")).toBe(true);
    expect(SerialPort.Binding).toEqual(binding);
  });

  it("should allow multiple ports to be opened", async () => {
    await Promise.all(mockMspDevices.map((port) => open(port)));
    expect(mockMspDevices.every((port) => isOpen(port))).toBe(true);
  });

  it("should throw an error when trying to open a port which doesn't respond with api version", () =>
    new Promise<void>((done) => {
      jest.useFakeTimers();

      open("/dev/non-msp-device")
        .then(() => {
          throw new Error("should not have resolved");
        })
        .catch((e) => {
          expect(e).toMatchSnapshot();
          done();
        });

      realSetTimeout(() => {
        jest.advanceTimersByTime(2500);
      }, 100);
    }));

  it("should provide a callback and close the connection when the connection closes", () => {
    return new Promise<void>((done) => {
      open("/dev/something", () => {
        expect(isOpen("/dev/something")).toBe(false);
        done();
      }).then(() => raw("/dev/something")?.close());
    });
  });

  it("should close the connection when an error occurs", () =>
    new Promise<void>((done) => {
      open("/dev/something", () => {
        expect(isOpen("/dev/something")).toBe(false);
        done();
      }).then(() =>
        raw("/dev/something")?.emit("error", new Error("Oh no some error"))
      );
    }));

  it("should close the connection if the device disconnects", () =>
    new Promise<void>((done) => {
      open("/dev/something", () => {
        expect(isOpen("/dev/something")).toBe(false);
        done();
      }).then(() => {
        MockBinding.reset();
        mockPorts
          .filter((port) => port !== "/dev/something")
          .forEach((path) => {
            MockBinding.createPort(path, {
              record: true,
            });
          });
      });
    }));

  it("should throw an error when port cannot be opened", async () => {
    await expect(open("/something/wrong")).rejects.toEqual(expect.any(Error));
    await expect(open("/something/wrong")).rejects.toMatchSnapshot();
  });

  it("should not allow a connection to be opened more than once", async () => {
    await open("/dev/something");
    await expect(open("/dev/something")).rejects.toEqual(expect.any(Error));
    await expect(open("/dev/something")).rejects.toMatchSnapshot();
  });
});

describe("connections", () => {
  it("should return a list of connected ports", async () => {
    expect(connections()).toEqual([]);
    await open("/dev/somethingelse");

    expect(connections()).toEqual(["/dev/somethingelse"]);

    await close("/dev/somethingelse");
    expect(connections()).toEqual([]);
  });
});

describe("close", () => {
  it("should close an open connection", async () => {
    const closeCallback = jest.fn();

    await open("/dev/something", closeCallback);
    await close("/dev/something");

    expect(isOpen("/dev/something")).toBe(false);
    expect(closeCallback).toHaveBeenCalled();
  });

  it("should ignore when connection is already closed", async () => {
    await expect(close("/dev/something")).resolves.toBeUndefined();
  });

  it("should cancel all pending executions", async () => {
    await open("/dev/something");

    const requests = Promise.all(
      [1, 2].map((v) =>
        execute("/dev/something", { code: v })
          .then(() => {
            throw new Error("Should have thrown an error");
          })
          .catch((e) => {
            expect(e).toMatchSnapshot();
          })
      )
    );
    await flushPromises();

    await close("/dev/something");

    return requests;
  });
});

describe("ports", () => {
  it("should list the available ports", async () => {
    expect(await ports()).toEqual(mockPorts);
  });
});

describe("execute", () => {
  it("should throw an error if the port is not open", async () => {
    await expect(execute("/dev/something", { code: 1 })).rejects.toEqual(
      expect.any(Error)
    );
    await expect(
      execute("/dev/something", { code: 1 })
    ).rejects.toMatchSnapshot();
  });

  it("should write the given command and data for v1 commands", async () => {
    await open("/dev/something");

    await execute("/dev/something", {
      code: 254,
      data: Buffer.from("This is a message"),
      timeout: 0,
    }).catch(() => {});

    expect(writtenData("/dev/something")).toEqual(
      encodeMessageV1(254, Buffer.from("This is a message"))
    );
  });

  it("should write the given command and data for v2 commands", async () => {
    await open("/dev/something");

    await execute("/dev/something", {
      code: 256,
      data: Buffer.from("This is a v2 message"),
      timeout: 0,
    }).catch(() => {});

    expect(writtenData("/dev/something")).toEqual(
      encodeMessageV2(256, Buffer.from("This is a v2 message"))
    );
  });

  it("should deduplicate requests, if the same request is already in flight", async () => {
    await open("/dev/something");

    await Promise.all([
      execute("/dev/something", {
        code: 254,
        data: Buffer.from("This is a message"),
        timeout: 0,
      }).catch(() => {}),
      execute("/dev/something", {
        code: 254,
        data: Buffer.from("This is a message"),
        timeout: 0,
      }).catch(() => {}),
    ]);

    expect(writtenData("/dev/something")).toEqual(
      encodeMessageV1(254, Buffer.from("This is a message"))
    );

    await execute("/dev/something", {
      code: 254,
      data: Buffer.from("This is a different request"),
      timeout: 0,
    }).catch(() => {});

    expect(writtenData("/dev/something")).toEqual(
      Buffer.concat([
        encodeMessageV1(254, Buffer.from("This is a message")),
        encodeMessageV1(254, Buffer.from("This is a different request")),
      ])
    );
  });

  it("should return response of the given request", async () => {
    await open("/dev/something");
    const execution = execute("/dev/something", {
      code: 108,
    });
    await flushPromises();

    // Reply the data in 2 parts to ensure that it can
    // handle data coming in chunks
    reply("/dev/something", Buffer.from([36, 77, 62, 6, 108]));
    await flushPromises();
    reply("/dev/something", Buffer.from([129, 0, 62, 1, 100, 1, 177]));

    const response = await execution;
    expect(Buffer.from(response.buffer)).toEqual(
      Buffer.from([129, 0, 62, 1, 100, 1])
    );
  });

  it("should ignore responses which are not related", async () => {
    await open("/dev/something");
    const execution = execute("/dev/something", {
      code: 108,
    });
    await flushPromises();

    // write some data which is not the response
    reply("/dev/something", Buffer.from([36, 77, 62, 0, 100, 177]));
    await flushPromises();
    reply(
      "/dev/something",
      Buffer.from([36, 77, 62, 6, 108, 129, 0, 62, 1, 100, 1, 177])
    );
    await flushPromises();

    const response = await execution;
    expect(Buffer.from(response.buffer)).toEqual(
      Buffer.from([129, 0, 62, 1, 100, 1])
    );
  });

  it("should return the same data for a duplicate request", async () => {
    await open("/dev/something");
    const execution1 = execute("/dev/something", {
      code: 108,
    });
    const execution2 = execute("/dev/something", {
      code: 108,
    });

    await flushPromises();
    reply(
      "/dev/something",
      Buffer.from([36, 77, 62, 6, 108, 129, 0, 62, 1, 100, 1, 177])
    );

    const response1 = await execution1;
    expect(Buffer.from(response1.buffer)).toEqual(
      Buffer.from([129, 0, 62, 1, 100, 1])
    );

    const response2 = await execution2;
    expect(Buffer.from(response2.buffer)).toEqual(
      Buffer.from([129, 0, 62, 1, 100, 1])
    );

    // Ensure they are their own objects
    expect(response1).not.toBe(response2);
  });

  it("should ignore malformed responses", async () => {
    await open("/dev/something");

    jest.useFakeTimers();
    const execution = execute("/dev/something", {
      code: 108,
    });

    // Invalid checksum
    reply(
      "/dev/something",
      Buffer.from([36, 77, 62, 6, 108, 129, 0, 62, 1, 100, 1, 232])
    );
    await flushPromises();

    jest.runAllTimers();
    // Should timeout from no response
    await expect(execution).rejects.toEqual(expect.any(Error));
    expect(packetErrors("/dev/something")).toBe(1);
  });

  it("should timeout after given timeout", async () => {
    await open("/dev/something");

    jest.useFakeTimers();

    let rejected: Error | undefined;
    execute("/dev/something", {
      code: 108,
      timeout: 500,
    }).catch((err: Error) => {
      rejected = err;
    });

    // Advancing to 499 should not timeout
    jest.advanceTimersByTime(499);
    await flushPromises();
    expect(rejected).toBeFalsy();

    // but 500 should
    jest.advanceTimersByTime(1);
    await flushPromises();

    expect(rejected).toBeTruthy();
    expect(rejected).toMatchSnapshot();
  });
});

describe("bytesRead", () => {
  it("should initialise at 0", async () => {
    await open("/dev/something");
    expect(bytesRead("/dev/something")).toEqual(0);
  });

  it("should count the number of bytes read from the device", async () => {
    const data = [36, 77, 62, 6, 108, 129, 0, 62, 1, 100, 1, 177];
    await open("/dev/something");
    reply("/dev/something", Buffer.from(data));
    await flushPromises();
    expect(bytesRead("/dev/something")).toEqual(data.length);
  });
});

describe("bytesWritten", () => {
  it("should initialise at 0", async () => {
    await open("/dev/something");
    expect(bytesWritten("/dev/something")).toEqual(0);
  });

  it("should count the number of bytes written to the device", async () => {
    await open("/dev/something");
    await execute("/dev/something", {
      code: 54,
      data: Buffer.from("This is a v1 message"),
      timeout: 0,
    }).catch(() => {});

    expect(bytesWritten("/dev/something")).toEqual(
      Buffer.byteLength(
        encodeMessageV1(54, Buffer.from("This is a v1 message"))
      )
    );
  });
});

describe("apiVersion", () => {
  it("should respond with the api version for the currently open port", async () => {
    await open("/dev/something");
    expect(apiVersion("/dev/something")).toEqual("1.40.0");
  });

  it("should respond with 0.0.0 when port is not open", () => {
    expect(apiVersion("/dev/something")).toEqual("0.0.0");
  });
});
