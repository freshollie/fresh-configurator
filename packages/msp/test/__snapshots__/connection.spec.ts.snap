// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`baudRate should throw an error if the port is not open 1`] = `"Port is not open"`;

exports[`close should cancel all pending executions 1`] = `[Error: Connection closed]`;

exports[`close should cancel all pending executions 2`] = `[Error: Connection closed]`;

exports[`execute should throw an error if the port is not open 1`] = `[Error: /dev/something is not open]`;

exports[`execute should timeout after given timeout 1`] = `[Error: Timed out during execution]`;

exports[`open should not allow a connection to be opened more than once 1`] = `[Error: /dev/something is already open]`;

exports[`open should throw an error when port cannot be opened 1`] = `[Error: Port does not exist - please call MockBinding.createPort('/something/wrong') first]`;

exports[`open should throw an error when trying to open a port which doesn't respond with api version 1`] = `[Error: Could not read MSP version from /dev/non-msp-device]`;
